### PR TITLE
More processing goodness, restore algs

### PR DESCRIPTION
**Pull request by nyalldawson**
- Clip alg allows mask in different CRS to clipped layer
- Port Extract by Expression to new API, allow saving non matching features to separate output

### CODE REVIEW

File: python/plugins/processing/algs/qgis/Clip.py [hunk 1]
```diff
@@ -84,7 +84,7 @@ def processAlgorithm(self, parameters, context, feedback):
 
         # first build up a list of clip geometries
         clip_geoms = []
-        for mask_feature in mask_source.getFeatures(QgsFeatureRequest().setSubsetOfAttributes([])):
+        for mask_feature in mask_source.getFeatures(QgsFeatureRequest().setSubsetOfAttributes([]).setDestinationCrs(feature_source.sourceCrs())):
```
Comment (Gustry) [inline python/plugins/processing/algs/qgis/Clip.py]: So it means we can do a clipping with 2 layers with different CRS ? Nice. It has to be written in the help of the algorithm.

**Comment by nirvn** [comment]
There's no way to extrapolate the CRS stuff to an higher level than algorithm? It'll become confusing for users to have a mixture of algorithms with varying support for multiple CRS inputs.

**Comment by nyalldawson** [comment]
> There's no way to extrapolate the CRS stuff to an higher level than algorithm? It'll become confusing for users to have a mixture of algorithms with varying support for multiple CRS inputs.

I don't think so - it needs logic in the alg to know which input's crs is the "correct" one. But it'd consider it a bug if any alg doesn't implement this now.

File: python/plugins/processing/algs/qgis/Clip.py (hunk 1 above)
Comment (nyalldawson) [inline python/plugins/processing/algs/qgis/Clip.py]: Yes - no more temporary reprojected layers just to clip two other layers!

**Comment by Gustry** [comment]
+1 with @nirvn.
It will be very confusing if people stop doing reprojection and if this feature is not implemented in ALL processing algs.

Can we add a label saying that this algorithm can work with multi-CRS inputs?

**Comment by nyalldawson** [comment]
> It will be very confusing if people stop doing reprojection and if this feature is not implemented in ALL processing algs.

> Can we add a label saying that this algorithm can work with multi-CRS inputs?

I disagree - I say we just block ANY algorithm which doesn't support them, and get the algorithm fixed before it's included. It's literally adding just a couple of characters for algs to add this support, it's no big effort.

**Comment by pcav** [comment]
Please document this, so also Processing plugin writer will use it.
Thanks, big improvement!

**Comment by Gustry** [comment]
Ok, so every algorithms are going to be updated? very nice! Thanks !

For some algorithms (the union for instance), we maybe need to inform the user about which CRS it will use (layer A or B).

**Comment by nirvn** [comment]
We have to consider that we'll have limited control over multi CRS inputs with 3rd party algs (eg grass7)

**Comment by nirvn** [comment]
We have to consider that we'll have limited control over multi CRS inputs with 3rd party algs (eg grass7)

**Comment by nyalldawson** [comment]
> We have to consider that we'll have limited control over multi CRS inputs with 3rd party algs (eg grass7)

I'd rather not cripple the native algorithms because of this. If it's an issue, it should be flagged in THOSE algorithms that they don't support mixed projections.

But it should be possible to just transparently reproject if necessary anyway...

**Comment by nirvn** [comment]
@nyalldawson , agreed, we should not weaken our native algorithms to match the limitation of 3rd party providers.

**Comment by nyalldawson** [comment]
Another quick benchtest - using the native c++ dissolve alg to dissolve a 50k complex polygon shapefile to a memory layer ~ 18 minutes. Dissolving the same layer in Postgis (using st_unaryunion(st_collect(geom)) ) took about 29 minutes :D

**Comment by gioman** [comment]
@nyalldawson 



> Another quick benchtest - using the native c++ dissolve alg to dissolve a 50k complex polygon shapefile to a memory layer ~ 18 minutes. Dissolving the same layer in Postgis (using st_unaryunion(st_collect(geom)) ) took about 29 minutes :D

trying dissolve this 

http://www.dgterritorio.pt/filedownload.aspx?schema=b511271f-54fe-4d21-9657-24580e9b7023&channel=6E9B5286-6507-4D17-8690-8A11FFE1FF35&content_id=DE5217EC-0969-498F-813F-F3381B4ECDB9&field=file&lang=pt&ver=1&filetype=rar

by "distrito" made my master (compiled on ubuntu 16.04) seg fault. No other messages in the console.

**Comment by gioman** [comment]
@nyalldawson actually dissolve seg faults with any input here.

**Comment by nyalldawson** [comment]
@gioman are you running a debug build? Can you get a backtrace?

**Comment by nyalldawson** [comment]
@gioman I can't reproduce. I've tried with your sample data and numerous others with no error. Can you share your workflow or a screencast?

**Comment by gioman** [comment]
@nyalldawson it could be my build. I will wait for a osgeo4w build that includes the dissolve tool and try there.

**Comment by gioman** [comment]
@nyalldawson 

> @gioman are you running a debug build? Can you get a backtrace?


I waited a Windows osgeor4w master build that included your new native c++ dissolve tool.

I tested it (on the same platform and conditions) against qgis 2.18.9 nightly, that also write debug output.

Using 
https://www.dropbox.com/s/ncj4ysh15xrnm45/caop.zip?dl=0

and dissolving by "distrito" the ogr2ogr based tool is 2/3 seconds faster than the new native one.


Using:
https://www.dropbox.com/s/qr8md49cy5u8vrf/clip.zip?dl=0

which is a clip of a much larger and complex layer, and dissolving by "aaaaa" (sorry for the column name), the ogr 2ogr based tool does the job in about 20 seconds.

The new native dissolve tool runs for a split second and outputs a few polygons in vastly incomplete result.

Trying to dissolve all polygons (and operation that the ogr2ogr based tool does with no issues) with the native tool it results in a python error and and infinite running hourglass.

```
An error has occurred while executing Python code: 

SystemError: returned a result with an error set 
Traceback (most recent call last):
  File "C:/OSGEO4~1/apps/qgis-dev/./python/plugins\processing\tools\dataobjects.py", line 84, in raise_error
    'Features with invalid geometries found. Please fix these geometries or specify the "Ignore invalid input features" flag'))
processing.core.GeoAlgorithmExecutionException.GeoAlgorithmExecutionException: 

 Message:
  | Features with invalid geometries found. Please fix these geometries or specify the "Ignore invalid input features" flag

 Stack:

    None

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:/OSGEO4~1/apps/qgis-dev/./python/plugins\processing\gui\AlgorithmDialog.py", line 246, in accept
    result = executeAlgorithm(self.alg, parameters, context, feedback)
  File "C:/OSGEO4~1/apps/qgis-dev/./python/plugins\processing\core\GeoAlgorithm.py", line 338, in executeAlgorithm
    result = alg.run(parameters, context, feedback)
SystemError:  returned a result with an error set
```

Interesting the massage

"Features with invalid geometries found. Please fix these geometries or specify the "Ignore invalid input features" flag"

because the option is not available in the tool. At this point I argue that -if it is really a problem of bad geometries- we miss a clear suggestion and workflow to allow the user solve this situation. In my point of view the option in such tools should not be one to "skip the invalid geometries" but to *fix* them.

cheers!

**Pull request merged**